### PR TITLE
[#143][#172] Add WithCheckDigit, refactor CreditCard

### LINKED ISSUE #143: Create WithCheckDigit
Modules like credit card, ISBN 8/10, VIN, and others have a checksum digit.  WithChecksumDigit will give an ability to replace / validate a particular digit with a digit created using a supplied algorithm.

### LINKED ISSUE #172: Refactor credit card module
depends on #143

### PR DESCRIPTION
**Pull request by skeate**
Closes #143 
Closes #172

### CODE REVIEW

File: scripts/generate-schemables.ts [hunk 1]
```diff
@@ -28,6 +28,17 @@ type SchemableCombinators = {
...
+type SchemableAugment = {
+  name: `With${string}`
+}
+
+const Augments: ReadonlyArray<SchemableAugment> = [
+  { name: 'WithBrand' },
+  { name: 'WithCheckDigit' },
+  { name: 'WithInvariant' },
+  { name: 'WithPattern' },
+]
```
Comment (skeate) [inline scripts/generate-schemables.ts:40]: I thought about adding `WithRefine` and `WithUnknownContainers` in here but they added a lot of extra logic when it's easier to just leave them hard coded

Comment (jacob-alford) [reply scripts/generate-schemables.ts:40]: I ended up adding them in my branch, so no worries!  It was indeed a huge pain in the butt

File: src/internal/DecoderBase.ts [hunk 2]
```diff
@@ -27,3 +30,31 @@ export const WithInvariant: WithInvariant2C<D.URI, unknown> = {
...
+      pipe(
+        dec.decode(s),
+        E.chain(
+          E.fromPredicate(
+            (s): s is Branded<string, CheckDigitVerified> =>
+              s[typeof location === 'number' ? location : location(s)] === algorithm(s),
+            s =>
+              D.error(
+                s,
+                replaceCharAt(
```
Comment (jacob-alford) [inline src/internal/DecoderBase.ts:50]: Why include the corrected string in the error?

File: src/schemata/string/CreditCard.ts [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+        PB.exactString('8'),
+        PB.characterClass(false, ['2', '8']),
+        PB.exactly(4)(PB.digit)
+      )
+    )
+  ),
+  PB.between(8, 11)(PB.digit)
+)
+
+const cc = PB.oneOf(visa, mastercard, amex, dinersClub, discover, jcb, rupay, unionPay)
```
Comment (jacob-alford) [inline src/schemata/string/CreditCard.ts:275]: This is incredibly comprehensive :+1:

File: tests/Decoder.test.ts [hunk 4]
```diff
@@ -4,7 +4,28 @@ import { Schemable } from '../src/Decoder'
...
+  it('decodes check digits', () => {
+    const CheckDigit = make(S => S.checkDigit(s => s[0] ?? '0', 1)(S.string))
+    const decode = interpreter(Schemable)(CheckDigit)
+    expect(decode.decode('01')).toEqual({
+      _tag: 'Left',
+      left: {
+        _tag: 'Of',
+        value: {
+          _tag: 'Leaf',
+          actual: '01',
```
Comment (jacob-alford) [inline tests/Decoder.test.ts:21]: Ahh I see now

**Review by jacob-alford — APPROVED** [review]
Looks great!  One question about including the checksum in the error.  Might make that something descriptive 🤷‍♂️ though normally it's just a name

File: tests/Decoder.test.ts (hunk 4 above)
Comment (skeate) [reply tests/Decoder.test.ts:21]: and with the actual decode error reporter or whatever it's called, this would look nicer

**Review by jacob-alford — APPROVED** [review]
LGTM

**Pull request merged**